File: flutter_pos_2013/lib/data/repository/payment_repository.dart
import '../../core/config/gateway_config.dart';
import '../../core/result/result.dart';
import '../../core/utils/date_util.dart';
import '../../core/utils/hmac_util.dart';
import '../../core/utils/id_generator.dart';
import '../../core/utils/encryption_util.dart';
import '../local/database_helper.dart';
import '../model/transaction_model.dart';
import '../remote/api_service.dart';

/// Payment Repository - Handles payment processing and syncing
class PaymentRepository {
  final DatabaseHelper _db;
  final ApiService _api;

  PaymentRepository({
    DatabaseHelper? databaseHelper,
    ApiService? apiService,
  })  : _db = databaseHelper ?? DatabaseHelper(),
        _api = apiService ?? ApiService();

  // ========== PAYMENT PROCESSING ==========

  /// Process a new card payment
  Future<PaymentResult> processPayment({
    required String cardNumber,
    required String expiry,
    String? cvv,
    required double amount,
  }) async {
    try {
      // Convert amount to minor units
      final amountMinor = (amount * 100).round();
      
      // Generate IDs
      final localTxnId = IdGenerator.generateLocalTxnId();
      final stan = await IdGenerator.generateStan();
      final timestamp = DateTime.now().millisecondsSinceEpoch;
      
      // Get card last 4 digits
      final cardLast4 = cardNumber.length >= 4 
          ? cardNumber.substring(cardNumber.length - 4) 
          : cardNumber;

      // Encrypt card data (if needed for storage)
      String? encryptedPan;
      try {
        final encUtil = EncryptionUtil();
        encryptedPan = encUtil.encrypt(cardNumber);
      } catch (e) {
        // Continue without encryption if it fails
        encryptedPan = null;
      }

      // Create transaction model
      final transaction = TransactionModel(
        id: localTxnId,
        localTxnId: localTxnId,
        stan: stan,
        amountMinor: amountMinor,
        encryptedPan: encryptedPan,
        cardLast4: cardLast4,
        cardExpiry: expiry,
        timestamp: timestamp,
        txnTimestamp: DateUtil.toIsoString(DateTime.now()),
        syncStatus: 'PENDING',
        synced: false,
      );

      // Save to local database
      await _db.insertTransaction(transaction);

      // Try to sync immediately
      try {
        final syncResult = await _syncTransaction(transaction);
        if (syncResult is Success<String?>) {
          return PaymentSuccess(
            localTxnId: localTxnId,
            stan: stan,
            amount: amount,
            settlementCode: syncResult.data,
            message: 'Payment processed successfully',
          );
        }
      } catch (e) {
        // Sync failed, keep as pending
      }

      // Return pending result
      return PaymentPending(
        localTxnId: localTxnId,
        stan: stan,
        amount: amount,
        message: 'Payment saved offline. Will sync when online.',
      );

    } catch (e) {
      return PaymentError('Failed to process payment: $e');
    }
  }

  /// Sync a single transaction
  Future<Result<String?>> _syncTransaction(TransactionModel transaction) async {
    try {
      // Update status to syncing
      await _db.updateTransactionStatus(
        transaction.localTxnId,
        status: 'SYNCING',
      );

      // Get config
      final merchantId = await GatewayConfig.merchantId;
      final terminalId = await GatewayConfig.terminalId;
      final secretKey = await GatewayConfig.secretKey;

      // Generate batch data
      final batchId = IdGenerator.generateBatchId();
      final nonce = HmacUtil.generateNonce();
      final timestamp = DateTime.now().millisecondsSinceEpoch;

      // Create transaction request
      final txnRequest = TransactionRequest(
        localTxnId: transaction.localTxnId,
        stan: transaction.stan,
        amountMinor: transaction.amountMinor,
        currency: transaction.currency,
        encryptedPan: transaction.encryptedPan,
        cardLast4: transaction.cardLast4,
        expiry: transaction.cardExpiry,
        txnType: transaction.txnType,
        entryMode: transaction.entryMode,
        txnTimestamp: transaction.txnTimestamp ?? DateUtil.toIsoString(DateTime.now()),
      );

      // Generate signature
      final signature = HmacUtil.generateSignature(
        protocolVersion: '201.3',
        merchantId: merchantId,
        terminalId: terminalId,
        batchId: batchId,
        timestamp: timestamp,
        nonce: nonce,
        transactionCount: 1,
        secretKey: secretKey,
      );

      // Create batch request
      final request = BatchUploadRequest(
        merchantId: merchantId,
        terminalId: terminalId,
        batchId: batchId,
        timestamp: timestamp,
        nonce: nonce,
        transactions: [txnRequest],
        signature: signature,
      );

      // Send to server
      final response = await _api.uploadBatch(request);

      if (response.success) {
        await _db.markAsSynced(
          transaction.localTxnId,
          response.settlementCode ?? '',
        );
        return Success(response.settlementCode);
      } else {
        await _db.updateTransactionStatus(
          transaction.localTxnId,
          status: 'FAILED',
          errorMessage: response.error ?? 'Unknown error',
        );
        return Failure(response.error ?? 'Sync failed');
      }

    } catch (e) {
      await _db.updateTransactionStatus(
        transaction.localTxnId,
        status: 'PENDING',
        errorMessage: e.toString(),
      );
      return Failure('Sync error: $e');
    }
  }

  /// Sync all pending transactions
  Future<SyncSummary> syncPendingTransactions() async {
    final pending = await _db.getPendingTransactions();
    
    int successCount = 0;
    int failedCount = 0;
    final settlementCodes = <String>[];

    for (final transaction in pending) {
      final result = await _syncTransaction(transaction);
      if (result is Success<String?>) {
        successCount++;
        if (result.data != null && result.data!.isNotEmpty) {
          settlementCodes.add(result.data!);
        }
      } else {
        failedCount++;
      }
    }

    return SyncSummary(
      total: pending.length,
      synced: successCount,
      failed: failedCount,
      settlementCodes: settlementCodes,
    );
  }

  // ========== REDEMPTION ==========

  /// Redeem a payment code
  Future<RedeemResult> redeemCode(String code, double amount) async {
    try {
      final merchantId = await GatewayConfig.merchantId;
      
      final request = RedeemRequest(
        code: code,
        amount: amount,
        merchantId: merchantId,
      );

      final response = await _api.redeemCode(request);

      if (response.success) {
        return RedeemSuccess(
          message: response.message ?? 'Payment redeemed successfully',
          reference: response.reference,
          settlementCode: response.settlementCode,
        );
      } else {
        return RedeemError(response.message ?? 'Redemption failed');
      }

    } catch (e) {
      return RedeemError('Error: $e');
    }
  }

  // ========== TRANSACTION QUERIES ==========

  /// Get all transactions
  Future<List<TransactionModel>> getAllTransactions() async {
    return await _db.getAllTransactions();
  }

  /// Get pending transactions
  Future<List<TransactionModel>> getPendingTransactions() async {
    return await _db.getPendingTransactions();
  }

  /// Get pending count
  Future<int> getPendingCount() async {
    return await _db.getPendingCount();
  }

  /// Get transaction by ID
  Future<TransactionModel?> getTransaction(String localTxnId) async {
    return await _db.getTransactionByLocalId(localTxnId);
  }

  // ========== CLEANUP ==========

  /// Clear old synced transactions
  Future<void> clearOldTransactions(int days) async {
    final millis = days * 24 * 60 * 60 * 1000;
    await _db.deleteOldSyncedTransactions(millis);
  }

  /// Delete all data
  Future<void> clearAllData() async {
    await _db.deleteAllTransactions();
    await _db.deleteAllOrders();
  }
}

File: android_pos_app/app/src/main/java/com/pos2013/offline/data/PaymentRepository.kt
package com.pos2013.offline.data

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.util.Log
import com.pos2013.offline.config.GatewayConfig
import com.pos2013.offline.data.api.*
import com.pos2013.offline.data.model.*
import com.pos2013.offline.utils.HmacUtil
import com.pos2013.offline.utils.IdGenerator
import com.pos2013.offline.utils.PanEncryptor
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.*

class PaymentRepository(
    private val context: Context
) {
    private val TAG = "PaymentRepository"
    private val storage = SimpleStorage(context)
    private val apiService = RetrofitClient.getApiService()
    private val isoFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
        timeZone = TimeZone.getTimeZone("UTC")
    }

    // Process a new payment
    suspend fun processPayment(
        cardNumber: String,
        cardExpiry: String,
        cardCvv: String?,
        amount: Double
    ): PaymentResult = withContext(Dispatchers.IO) {
        try {
            // Use BigDecimal to avoid floating point rounding issues when converting to minor units.
            val amountMinor = try {
                java.math.BigDecimal.valueOf(amount)
                    .multiply(java.math.BigDecimal.valueOf(100))
                    .setScale(0, java.math.RoundingMode.HALF_UP)
                    .intValueExact()
            } catch (e: ArithmeticException) {
                Int.MAX_VALUE
            }
            val localTxnId = IdGenerator.generateLocalTxnId()
            val stan = IdGenerator.generateStan(context)
            val timestamp = System.currentTimeMillis()

            // REAL WORLD SECURITY: Encrypt PAN before saving to storage
            val encryptedPan = try {
                PanEncryptor.encrypt(context, cardNumber)
            } catch (e: Exception) {
                Log.e(TAG, "Encryption failed", e)
                null
            }
            
            val cardLast4 = cardNumber.takeLast(4)

            val transaction = StoredTransaction(
                localTxnId = localTxnId,
                stan = stan,
                amountMinor = amountMinor,
                cardLast4 = cardLast4,
                encryptedPan = encryptedPan,
                cardExpiry = cardExpiry,
                timestamp = timestamp,
                syncStatus = "PENDING",
                synced = false
            )

            // Save to storage immediately
            storage.saveTransaction(transaction)

            // Try to sync immediately if online
            if (isNetworkAvailable()) {
                val syncResult = syncTransaction(transaction)

                if (syncResult is SyncResult.Success) {
                    PaymentResult.Success(
                        localTxnId = localTxnId,
                        stan = stan,
                        amount = amount,
                        settlementCode = syncResult.settlementCode,
                        message = "Payment processed successfully"
                    )
                } else {
                    PaymentResult.Pending(
                        localTxnId = localTxnId,
                        stan = stan,
                        amount = amount,
                        message = "Saved offline. Will sync when online."
                    )
                }
            } else {
                PaymentResult.Pending(
                    localTxnId = localTxnId,
                    stan = stan,
                    amount = amount,
                    message = "No internet. Payment saved and will sync when online."
                )
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error processing payment", e)
            PaymentResult.Error("Failed to process: ${e.message}")
        }
    }

    // Sync a single transaction
    private suspend fun syncTransaction(transaction: StoredTransaction): SyncResult {
        return try {
            storage.updateStatus(transaction.localTxnId, "SYNCING")

            val batchId = IdGenerator.generateBatchId()
            val nonce = HmacUtil.generateNonce()
            val timestamp = System.currentTimeMillis()

            val txnRequest = TransactionRequest(
                localTxnId = transaction.localTxnId,
                stan = transaction.stan,
                amountMinor = transaction.amountMinor,
                currency = transaction.currency,
                encryptedPan = transaction.encryptedPan,
                cardLast4 = transaction.cardLast4,
                expiry = transaction.cardExpiry,
                txnType = "SALE",
                entryMode = "MANUAL",
                txnTimestamp = isoFormat.format(Date(transaction.timestamp))
            )

            val signature = HmacUtil.generateSignature(
                protocolVersion = "201.3",
                merchantId = GatewayConfig.MERCHANT_ID,
                terminalId = GatewayConfig.TERMINAL_ID,
                batchId = batchId,
                timestamp = timestamp,
                nonce = nonce,
                transactionCount = 1
            )

            val request = BatchUploadRequest(
                protocolVersion = "201.3",
                merchantId = GatewayConfig.MERCHANT_ID,
                terminalId = GatewayConfig.TERMINAL_ID,
                batchId = batchId,
                timestamp = timestamp,
                nonce = nonce,
                transactions = listOf(txnRequest),
                signature = signature
            )

            val response = apiService.uploadBatch(signature, request)

            if (response.isSuccessful) {
                val body = response.body()
                if (body?.success == true) {
                    storage.markAsSynced(transaction.localTxnId, body.settlementCode ?: "")
                    SyncResult.Success(body.settlementCode)
                } else {
                    val error = body?.error ?: "Unknown error"
                    handleSyncFailure(transaction, error)
                    SyncResult.Failed(error)
                }
            } else {
                val error = "Server error: ${response.code()}"
                handleSyncFailure(transaction, error)
                SyncResult.Failed(error)
            }
        } catch (e: Exception) {
            handleSyncFailure(transaction, e.message ?: "Sync failed")
            SyncResult.Failed(e.message ?: "Sync failed")
        }
    }

    private suspend fun handleSyncFailure(transaction: StoredTransaction, error: String) {
        storage.updateStatus(transaction.localTxnId, "FAILED", error)
    }

    suspend fun syncPendingTransactions(): SyncSummary = withContext(Dispatchers.IO) {
        val pending = storage.getPendingTransactions()
        var successCount = 0
        var failedCount = 0
        val settlementCodes = mutableListOf<String>()

        for (transaction in pending) {
            when (val result = syncTransaction(transaction)) {
                is SyncResult.Success -> {
                    successCount++
                    result.settlementCode?.let { settlementCodes.add(it) }
                }
                is SyncResult.Failed -> {
                    failedCount++
                }
            }
            delay(100)
        }

        SyncSummary(
            total = pending.size,
            synced = successCount,
            failed = failedCount,
            settlementCodes = settlementCodes
        )
    }

    // Redeem code
    suspend fun redeemCode(code: String, amount: Double): RedeemResult = withContext(Dispatchers.IO) {
        try {
            val request = RedeemRequest(
                code = code,
                amount = amount,
                merchantId = GatewayConfig.MERCHANT_ID
            )
            val response = apiService.redeemPaymentCode(request)
            if (response.isSuccessful) {
                val body = response.body()
                if (body?.success == true) {
                    RedeemResult.Success(
                        message = body.message ?: "Payment successful",
                        reference = body.reference,
                        settlementCode = body.settlementCode
                    )
                } else {
                    RedeemResult.Error(body?.message ?: "Redemption failed")
                }
            } else {
                RedeemResult.Error("Server error: ${response.code()}")
            }
        } catch (e: Exception) {
            RedeemResult.Error("Error: ${e.message}")
        }
    }

    // Verify credentials
    suspend fun verifyCredentials(merchantId: String, terminalId: String, secretKey: String): Boolean = withContext(Dispatchers.IO) {
        try {
            val request = VerifyRequest(merchantId, terminalId, secretKey)
            val response = apiService.verifyCredentials(request)
            response.isSuccessful && response.body()?.valid == true
        } catch (e: Exception) {
            false
        }
    }

    fun getPendingCount(): Int = storage.getPendingCount()

    fun getPendingTransactions(): List<StoredTransaction> = storage.getPendingTransactions()

    fun clearOldTransactions(olderThanMillis: Long) {
        storage.clearOldSynced(olderThanMillis)
    }

    fun isNetworkAvailable(): Boolean {
        val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val network = connectivityManager.activeNetwork ?: return false
        val capabilities = connectivityManager.getNetworkCapabilities(network) ?: return false
        return capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
    }
}

sealed class SyncResult {
    data class Success(val settlementCode: String?) : SyncResult()
    data class Failed(val error: String) : SyncResult()
}
